### PR TITLE
Add Python test coverage to codecov

### PR DESCRIPTION
**Pull request by shubhank-saxena**
`Tried to create hashes as well for the above stated package. Not sure what is meant by " whole hierarchy of packages with their hashes"

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -44,13 +44,15 @@ matrix:
       language: minimal
       install:
         - docker-compose build
+
```
Comment (armenzg) [inline .travis.yml]: Please remove the extra white lines.

File: ui/perfherder/alerts/Assignee.jsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ export default class Assignee extends React.Component {
     this.state = {
       assigneeUsername,
       inEditMode: false,
-      newAssigneeUsername: null,
+      newAssigneeUsername: "",
```
Comment (armenzg) [inline ui/perfherder/alerts/Assignee.jsx]: I think you need to pull the latest changes from the upstream repo and then `git merge master`.
I assume this change is not intended.

File: requirements/dev.txt [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Dependencies needed only for development/testing.
 
+codecov==2.0.15 \
```
Comment (armenzg) [inline requirements/dev.txt]: It seems that `codecov` depends on the package `coverage`.
That means you will need to also include that package (since it is part of the hierarchy) and its associated hash.

**Comment by armenzg** [comment]
You probably saw something like this on your machine (this is the output from Travis):
```
In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:

    coverage from https://files.pythonhosted.org/packages/9a/08/7fa92679a903cac0e01a7fee2cadd92089d96a3002c34e8f2295b80d6b68/coverage-5.0.3-cp37-cp37m-manylinux1_x86_64.whl#sha256=d649dc0bcace6fcdb446ae02b98798a856593b19b637c1b9af8edadf2b150bea (from codecov==2.0.15->-r requirements/dev.txt (line 3))
```
Notice it mentions the coverage package and the hash for it.

**Comment by shubhank-saxena** [comment]
@armenzg Fixed the issues as stated. Some minor spacing issues left in assignee.jsx. Will do that in next commit.

**Comment by shubhank-saxena** [comment]
Tried the following as you suggested - 

docker-compose run backend bash
pytest tests/ --runslow --ignore=tests/selenium/
pytest --cov=./ --cov-report=xml
codecov
     
Screenshot from 2020-01-24 04-46-50

I think we need a automatic script that is to be run as soon as we enter the bash mode because the commands are not running.Basically yaml file only handles the commands that are to be run on travis bash. For running commands inside a bash of our own, we need some sort of scripts. Moreover, the CI test will never complete as it doesn't have an exit. So we need to have an exit point too.

**Comment by shubhank-saxena** [comment]
@armenzg @klahnakoski the Travis build is being suspended on commit **bf9d7cd**. Can you please explain how we use runtests.sh and .travis.yml to run docker bash commands?

**Comment by shubhank-saxena** [comment]
@armenzg the Travis build is being suspended on commit **bf9d7cd**. Can you please explain how we use runtests.sh and .travis.yml to run docker bash commands?

**Comment by klahnakoski** [comment]
@shubhank-saxena   The line 

    docker-compose run backend bash

opens a bash instance, which is interactive, and waiting for human input.  It will seem to be suspended

**Comment by klahnakoski** [comment]
@shubhank-saxena   if you want to run multiple commands, and keep the side effects of each, then you must concatenate them with `&&`

**Comment by shubhank-saxena** [comment]
> @shubhank-saxena The line
> 
> ```
> docker-compose run backend bash
> ```
> 
> opens a bash instance, which is interactive, and waiting for human input. It will seem to be suspended

Okay. I thought the tests were supposed to be completely automatic. Then I will run the codecov tests manually on the travis instance. Thank you

**Comment by klahnakoski** [comment]
@shubhank-saxena   Yes, it is a bit confusing; the travis docker image is building a treeherder docker image, and the tests are run inside that second image.

**Comment by shubhank-saxena** [comment]
> @shubhank-saxena Yes, it is a bit confusing; the travis docker image is building a treeherder docker image, and the tests are run inside that second image.

Makes sense now!

**Comment by shubhank-saxena** [comment]
@klahnakoski so we have to run tests on local machine, run codecov on it, and then push it so it can go through Travis pipeline?

**Comment by klahnakoski** [comment]
@shubhank-saxena  The problem is with this line in travis.yml file:

    - docker-compose run backend bash

The first part of the line starts a new docker image (`docker-compose run backend`), the rest is the command used on the fresh image.  Your challenge is to compose a single line command the will run the tests and collect coverage.  You may have been close with these lines:

        - docker-compose run backend bash pytest tests/ --runslow --ignore=tests/selenium/
        - docker-compose run backend bash pytest --cov=./ --cov-report=xml
        - docker-compose run backend bash codecov         

...except for the fact this is multiple lines.

You can run multiple commands on a single line by concatenating them with `&&`:

        - docker-compose run backend pytest tests/ --runslow --ignore=tests/selenium/ && pytest --cov=./ --cov-report=xml  && codecov 

... but that may not work because of escaping and command arguments getting mixed up.  Notice i am not using `bash`; I am simply running a command

You can make a shell script, called `/tests/coverage.sh` and put all your commands in there.  You may need to set the execution flag.  Then you can run your command with a single-line in the travis.yml file

        - docker-compose run backend ./tests/coverage.sh 

I am not sure any of these particular commands actually work; I may have missed spaces and added typos.  You must experiment with exact phrasing. You should be able to start a local `backend` instance and use it to experiment with the instructions necessary to collect-and-submit coverage.  

In one command window:

    docker-compose up --build

In a separate command window:

    docker-compose run backend bash

Notice the use of `bash` this time: Which is only for humans. If you can not get the above two lines to work, then I can help you with that.

**Comment by shubhank-saxena** [comment]
@klahnakoski finally after 4 days, successfully got the codeov coverage of treeherder backed (sorry for excitement!) . Now will try with the coverage.sh method as you suggested.
Screenshot from 2020-01-27 16-11-46

**Comment by klahnakoski** [comment]
@shubhank-saxena good work!  Working with Docker the first few times takes effort.  It is good you are sticking with it.

**Comment by shubhank-saxena** [comment]
@klahnakoski the task is done!
1. Added coverage.sh run script in travis.yml
2. Created ./tests/coverage.sh 
One more PR would be required to just change the token number of the required output, as currently it pushes codecov to my repo

**Comment by shubhank-saxena** [comment]
Screenshot from 2020-01-28 18-02-28

File: .travis.yml [hunk 4]
```diff
@@ -50,8 +50,8 @@ matrix:
         # value so the tests pass. The real environment variable will be checked during deployment.
         - docker-compose run -e SITE_URL=https://treeherder.dev -e TREEHERDER_DEBUG=False backend python -bb ./manage.py check --deploy --fail-level WARNING
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/ --runslow --ignore=tests/selenium/
-
+        - docker-compose run backend bash ./tests/coverage.sh
+
```
Comment (klahnakoski) [inline .travis.yml]: I see that you used `bash` and I misled you.  Good job for working around my mistakes!

File: coverage.xml [hunk 5]
```diff
@@ -0,0 +1,14159 @@
+<?xml version="1.0" ?>
```
Comment (klahnakoski) [inline coverage.xml]: Should this file be in `.gitignore`?

Comment (shubhank-saxena) [reply coverage.xml]: Actually codecov after running tests, saves a file in .xml version. And that file is analysed and pushed over the codecov.
```text
pytest --cov=./ --cov-report=xml
```
This is the line responsible for that

File: .travis.yml (hunk 4 above)
Comment (shubhank-saxena) [reply .travis.yml]: @klahnakoski thank you! If you can provide me the token id of the repo, I can make the changes

File: coverage.xml (hunk 5 above)
Comment (klahnakoski) [reply coverage.xml]: Then remove this file and add it to `.gitignore`.  The codecov will make a new one after running tests, submit it, and we can ignore it.

Comment (shubhank-saxena) [reply coverage.xml]: Okay!

Comment (shubhank-saxena) [reply coverage.xml]: Done!

File: .travis.yml [hunk 6]
```diff
@@ -50,8 +50,8 @@ matrix:
         # value so the tests pass. The real environment variable will be checked during deployment.
         - docker-compose run -e SITE_URL=https://treeherder.dev -e TREEHERDER_DEBUG=False backend python -bb ./manage.py check --deploy --fail-level WARNING
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/ --runslow --ignore=tests/selenium/
-
+        - docker-compose run backend bash ./tests/coverage.sh
```
Comment (armenzg) [inline .travis.yml]: I believe you don't need to use `bash` in that line. Let me know if I'm wrong.

File: .gitignore [hunk 7]
```diff
@@ -35,3 +35,4 @@ celerybeat-schedule
 
 # Ignore coverage results
 coverage
+coverage.xml
```
Comment (armenzg) [inline .gitignore]: Add `.coverage` in here. Remove the `.coverage` file that is included in this PR.

File: .travis.yml [hunk 8]
```diff
@@ -74,3 +74,4 @@ notifications:
   email:
     on_success: never
     on_failure: always
+
```
Comment (armenzg) [inline .travis.yml]: Remove extra white line unless there's a reason for it.

File: tests/coverage.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+pytest tests/ --runslow --ignore=tests=selenium/
```
Comment (armenzg) [inline tests/coverage.sh]: Change this line for this: `exec "$@"`

File: .travis.yml (hunk 6 above)
Comment (armenzg) [reply .travis.yml]: I want to take advantage of `exec "$@"` in the coverage script.

You can then do this:
```
docker-compose run backend ./tests/coverage.sh ./runtests.sh
docker-compose run backend ./tests/coverage.sh pytest tests/ --runslow --ignore=tests=selenium/
docker-compose run backend ./tests/coverage.sh python -bb -m pytest tests/selenium/
```

**Comment by armenzg** [comment]
Good job @shubhank-saxena !

Do you have a link to your codecov account?

Run this command to validate the formatting of your Travis yml file: `yarn format:check`

**Comment by armenzg** [comment]
I found the link:
https://codecov.io/gh/shubhank-saxena/treeherder

It is a good idea to create a different branch instead of using `master` for development.
You can configure codecov to receive results from your development branch.

**Comment by armenzg** [comment]
I also can't notice the code coverage results for the UI:
image

File: tests/coverage.sh (hunk 9 above)
Comment (armenzg) [reply tests/coverage.sh]: I thought of this a bit better. Please remove this script.

File: .travis.yml (hunk 6 above)
Comment (armenzg) [reply .travis.yml]: Instead of taking advantage of the `exec` trick and the script. Use `after_success: pytest --cov=./ --cov-report=xml && codecov -t 995256be-3a1f-4d10-b181-e24f88603b6b`

Comment (shubhank-saxena) [reply .travis.yml]: > I believe you don't need to use `bash` in that line. Let me know if I'm wrong.

Sir, if we don't use bash, ./docker/entrypoint.sh is denying the permission to execute

Comment (shubhank-saxena) [reply .travis.yml]: > Instead of taking advantage of the `exec` trick and the script. Use `after_success: pytest --cov=./ --cov-report=xml && codecov -t 995256be-3a1f-4d10-b181-e24f88603b6b`

The issue with this is that there are two docker instances. One for Travis and one for our backend. Now the issue with after success is that it will try to find the xml file in the travis docker file, which it won't find. That's why, I tried to run and push the tests coverage in the inner docker container itself.

**Comment by shubhank-saxena** [comment]
> I also can't notice the code coverage results for the UI:
> image

Sir, I just handled the backend reporting right now. Let me check for pre-configured endpoints for ui part too. Let me check them and will report them all together.

File: .travis.yml (hunk 6 above)
Comment (armenzg) [reply .travis.yml]: Oh dear. This is why we need the GSoC project.

Use the `exec` approach I first mentioned.

Comment (shubhank-saxena) [reply .travis.yml]: > I want to take advantage of `exec "$@"` in the coverage script.
> 
> You can then do this:
> 
> ```
> docker-compose run backend ./tests/coverage.sh ./runtests.sh
> docker-compose run backend ./tests/coverage.sh pytest tests/ --runslow --ignore=tests=selenium/
> docker-compose run backend ./tests/coverage.sh python -bb -m pytest tests/selenium/
> ```

@armenzg I am not able to understand the outcome goal here. Can you please brief this a little more?

Comment (armenzg) [reply .travis.yml]: You will need to change the following lines:
https://github.com/mozilla/treeherder/blob/master/.travis.yml#L41
https://github.com/mozilla/treeherder/blob/master/.travis.yml#L53
https://github.com/mozilla/treeherder/blob/master/.travis.yml#L71

We run the Python tests in 3 batches and all will have to generate and upload coverage information.

**Comment by shubhank-saxena** [comment]
@armenzg @klahnakoski this is my understanding of the entire workflow of the tests. Where does the expected fit in? 
Workflow understanding

File: .travis.yml (hunk 6 above)
Comment (shubhank-saxena) [reply .travis.yml]: > You will need to change the following lines:
> https://github.com/mozilla/treeherder/blob/master/.travis.yml#L41
> https://github.com/mozilla/treeherder/blob/master/.travis.yml#L53
> https://github.com/mozilla/treeherder/blob/master/.travis.yml#L71
> 
> We run the Python tests in 3 batches and all will have to generate and upload coverage information.

Understood. Need a standard coverage file which should be passed as a parameter to every test that we are conducting. For that, need to use ```exec```

**Comment by shubhank-saxena** [comment]
Commit 3614c93 still needs some modifications. For some reasons, && is not working in the exec commands as I am trying to implement this in the coverage.sh - 
```exec pytest --cov=./ --cov-report=xml ; codecov -t 995256be-3a1f-4d10-b181-e24f88603b6b```
This is just creating the xml file and not reporting it to codecov

**Comment by armenzg** [comment]
Could you try this pattern? You would not need the `coverage.sh` script.

```yml
- install
  - docker-compose build
  - pip install codecov // <-- This means you won't need to specify codecov in dev.txt
- script
   - docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml <path/to/test> <other_options>"
- after_success:
  - codecov -t 995256be-3a1f-4d10-b181-e24f88603b6b
```

The following response is for your learning. If the above works we won't need to think about the following.

> @armenzg commit [3614c93](https://github.com/mozilla/treeherder/commit/3614c93df81602c89524130cda599c179e31a9c4) still needs some modifications. For some reasons, && is not working in the exec commands as I am trying to implement this in the coverage.sh -

Handling `&&` is actually handling more than one option after `bash`. You need to use `-c`. For instance, `bash -c "echo 'Hello world'` as this won't work `bash echo "Hello world"`.

> `exec pytest --cov=./ --cov-report=xml && codecov -t 995256be-3a1f-4d10-b181-e24f88603b6b`

The idea is this:

Contents of `script.sh`
```
#/bin/bash
exec "$@"
```
I can then call it like `script.sh echo "Hello world!"` and it will output the execution of the arguments to the script. It is a trick to execute text (a set of words that are passed as arguments) as a command. That way you can create a common script that can execute a command.

**Comment by shubhank-saxena** [comment]
@armenzg done! Kindly check and do let me know if any further changes are required

File: .travis.yml [hunk 10]
```diff
@@ -37,20 +37,26 @@ matrix:
...
         - docker-compose build
+        - pip install codecov --user
       script:
         # Several security features in settings.py (eg setting HSTS headers) are conditional on
         # 'https://' being in the site URL. In addition, we override the test environment's debug
         # value so the tests pass. The real environment variable will be checked during deployment.
         - docker-compose run -e SITE_URL=https://treeherder.dev -e TREEHERDER_DEBUG=False backend python -bb ./manage.py check --deploy --fail-level WARNING
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/ --runslow --ignore=tests/selenium/
+        - sudo docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml" python -bb -m pytests tests/ --runslow --ignore=tests/selenium
```
Comment (armenzg) [inline .travis.yml]: Change it to:
```bash
docker-compose run backend bash -c "pytest --cov-report=xml tests/ --runslow --ignore=tests/selenium"
```
Notice that I've changed the 2nd quote, I've removed "sudo" and dropped the `-bb` (this change is unrelated to your patch; it was simply overdue).

File: .travis.yml [hunk 11]
```diff
@@ -37,20 +37,26 @@ matrix:
       language: minimal
       install:
         - docker-compose build
+        - pip install codecov --user
```
Comment (armenzg) [inline .travis.yml]: I don't believe `--user` is necessary'. Am I correct?

File: .travis.yml [hunk 12]
```diff
@@ -37,20 +37,26 @@ matrix:
...
       script:
         # Several security features in settings.py (eg setting HSTS headers) are conditional on
         # 'https://' being in the site URL. In addition, we override the test environment's debug
         # value so the tests pass. The real environment variable will be checked during deployment.
         - docker-compose run -e SITE_URL=https://treeherder.dev -e TREEHERDER_DEBUG=False backend python -bb ./manage.py check --deploy --fail-level WARNING
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/ --runslow --ignore=tests/selenium/
+        - sudo docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml" python -bb -m pytests tests/ --runslow --ignore=tests/selenium
+      after_success :
+        - codecov -f coverage.xml -t 995256be-3a1f-4d10-b181-e24f88603b6b
```
Comment (armenzg) [inline .travis.yml]: What do I do to generate a key for the Treeherder codecov account?

File: initialize_data.sh [hunk 13]
```diff
@@ -8,4 +8,4 @@ echo '-----> Running Django migrations and loading reference data'
 echo '-----> Initialize SETA'
```
Comment (armenzg) [inline initialize_data.sh]: I believe you can remove this file now.

File: .gitignore [hunk 14]
```diff
@@ -35,3 +35,5 @@ celerybeat-schedule
 
 # Ignore coverage results
 coverage
+coverage.xml 
+.coverage
```
Comment (armenzg) [inline .gitignore]: This comment is not for this file but the one above:
image

Could you please remove the file `.coverage`?

A good practice for you to have is to review your own PR before asking for review. That might have helped you noticed that such file is part of your PR.

File: .travis.yml [hunk 15]
```diff
@@ -37,20 +37,26 @@ matrix:
       language: minimal
       install:
         - docker-compose build
+        - pip install codecov --user
       script:
-        - docker-compose run backend ./runtests.sh
+        - docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml" ./runtests.sh
```
Comment (armenzg) [inline .travis.yml]: Undo this change.

Instead change https://github.com/mozilla/treeherder/blob/master/runtests.sh#L30 to `pytest --cov-report=xml tests/`. Ignore the `python -bb -m pytests` you will see there.

File: .travis.yml [hunk 16]
```diff
@@ -63,12 +69,15 @@ matrix:
...
       install:
         - yarn install
+        - pip install codecov --user
       before_script:
         # Run in `before_script` to prevent the Selenium tests from running if the UI build fails.
         - yarn build
       script:
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/selenium/
+        - docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml" python -bb -m pytest tests/selenium/
```
Comment (armenzg) [inline .travis.yml]: This:
```
docker-compose run backend bash -c "pytest --cov-report=xml tests/selenium/"
```

File: runtests.sh [hunk 17]
```diff
@@ -27,4 +27,4 @@ echo "Running Django system checks"
 SITE_URL="https://treeherder.dev" TREEHERDER_DEBUG="False" python -bb ./manage.py check --deploy --fail-level WARNING
 
 echo "Running Python tests"
-python -bb -m pytest tests/
\ No newline at end of file
+python -bb -m pytest tests/
```
Comment (armenzg) [inline runtests.sh]: Change this to `pytest --cov-report=xml tests/`.

**Comment by shubhank-saxena** [comment]
> Good perseverance and happy to see the results published!
> 
> In general, place closer attention to details. Take the time to look closely at where double quotes are placed. Try the commands locally. Review the PR as if you were the reviewer. These are good practices to have.
> 
> Are the `ui` tests publishing coverage code for you? This screenshot compares to our current codecov dashboard:
> image
> 
> Review the output of the Travis Js tests. There's a formatting command that is showing some errors. Try the command and look for a parameter that fixes it.

Thank you so much. I will definitely keep in mind all of the points that you stated. 
1. Yes. I did try the commands locally and worked. But let me correct the double quote part.
2. No. for this -t token, I am using just the backend. Let me add the token for the UI too so that I can get the entire coverage of the codebase.
3. I think for the JS part, I did downloaded the file from origin/master as I messed up during the changes. Let me check where it is wrong.

File: .travis.yml (hunk 10 above)
Comment (shubhank-saxena) [reply .travis.yml]: Okay. Will make the required changes

File: .travis.yml (hunk 11 above)
Comment (shubhank-saxena) [reply .travis.yml]: Locally it wasn't needed. But when I tried my build, it failed stating that I need to put --user for providing privileges to installation.

File: .travis.yml (hunk 12 above)
Comment (shubhank-saxena) [reply .travis.yml]: You can click on the settings tab (right top corner) on the Treeherder codecov account page. You will find the key over there.

File: initialize_data.sh (hunk 13 above)
Comment (shubhank-saxena) [reply initialize_data.sh]: Okay. Will remove this file

File: .gitignore (hunk 14 above)
Comment (shubhank-saxena) [reply .gitignore]: Really sorry, sir. Will keep this in mind

File: runtests.sh (hunk 17 above)
Comment (shubhank-saxena) [reply runtests.sh]: Okay!

File: .travis.yml (hunk 16 above)
Comment (shubhank-saxena) [reply .travis.yml]: Okay!

File: .travis.yml (hunk 10 above)
Comment (shubhank-saxena) [reply .travis.yml]: If I am applying ```" "``` across the entire statement, it's throwing me the file not found error. That's why I wrote it as ```docker-compose run backend bash -c "pytest --cov-report=xml" tests/ --runslow --ignore=tests/selenium```
Screenshot from 2020-01-31 23-22-49

File: .travis.yml (hunk 12 above)
Comment (armenzg) [reply .travis.yml]: Thank you!

I noticed this information:
> Note: Token not required for public repositories uploading from Travis, CircleCI or AppVeyor.

That probably means that you don't need to specify it.

**Comment by armenzg** [comment]
For the record, you don't have to worry about deleting comments. It's OK to make a comment and then rectify yourself. I do that all the time even to this day.

Also please let me know if in any way I can convey feedback in manner that does not in any way discourage you. I tend to be speak straight to the point and I don't mean ill. You're working hard and you're trying well.

If I say anything wrong (technically or simple communication) please feel welcome to correct me. You can also email me privately if you prefer to.

Have a good weekend!

**Comment by shubhank-saxena** [comment]
> For the record, you don't have to worry about deleting comments. It's OK to make a comment and then rectify yourself. I do that all the time even to this day.
> 
> Also please let me know if in any way I can convey feedback in manner that does not in any way discourage you. I tend to be speak straight to the point and I don't mean ill. You're working hard and you're trying well.
> 
> If I say anything wrong (technically or simple communication) please feel welcome to correct me. You can also email me privately if you prefer to.
> 
> Have a good weekend!

Sorry sir. It's just that I posted an issue, and then I got the solution on my own. So I deleted it. Will keep in mind to edit it onwards. And I am not at all discouraged. Infcat it has motivated me to work 6+ hours in a shot!

**Comment by shubhank-saxena** [comment]
Yes! Finally rectified everything that you mentioned. Got all greens on my Travis Build( the total feeling of satisfaction) and I have covered the complete code coverage (including UI)
Travis Build - https://travis-ci.org/shubhank-saxena/treeherder/builds/644507647
Codecov - https://codecov.io/gh/shubhank-saxena/treeherder

**Comment by armenzg** [comment]
> > For the record, you don't have to worry about deleting comments. It's OK to make a comment and then rectify yourself. I do that all the time even to this day.
> > Also please let me know if in any way I can convey feedback in manner that does not in any way discourage you. I tend to be speak straight to the point and I don't mean ill. You're working hard and you're trying well.
> > If I say anything wrong (technically or simple communication) please feel welcome to correct me. You can also email me privately if you prefer to.
> > Have a good weekend!
> 
> Sorry sir. It's just that I posted an issue, and then I got the solution on my own. So I deleted it. Will keep in mind to edit it onwards. And I am not at all discouraged. Infcat it has motivated me to work 6+ hours in a shot!

That's hard work! 👍 If you have other responsibilities don't feel like you need to rush anything. I encourage you to take your time but whatever you feel comfortable with.

There's a merge conflict:
image

**Comment by shubhank-saxena** [comment]
> > > For the record, you don't have to worry about deleting comments. It's OK to make a comment and then rectify yourself. I do that all the time even to this day.
> > > Also please let me know if in any way I can convey feedback in manner that does not in any way discourage you. I tend to be speak straight to the point and I don't mean ill. You're working hard and you're trying well.
> > > If I say anything wrong (technically or simple communication) please feel welcome to correct me. You can also email me privately if you prefer to.
> > > Have a good weekend!
> > 
> > 
> > Sorry sir. It's just that I posted an issue, and then I got the solution on my own. So I deleted it. Will keep in mind to edit it onwards. And I am not at all discouraged. Infcat it has motivated me to work 6+ hours in a shot!
> 
> That's hard work! +1 If you have other responsibilities don't feel like you need to rush anything. I encourage you to take your time but whatever you feel comfortable with.
> 
> There's a merge conflict:
> image
@armenzg 
Sure. Thanks a lot. I have resolved the merge conflict and have rectified everything!

File: .travis.yml [hunk 18]
```diff
@@ -37,20 +37,26 @@ matrix:
...
         - docker-compose build
+        - pip install codecov --user
       script:
         # Several security features in settings.py (eg setting HSTS headers) are conditional on
         # 'https://' being in the site URL. In addition, we override the test environment's debug
         # value so the tests pass. The real environment variable will be checked during deployment.
         - docker-compose run -e SITE_URL=https://treeherder.dev -e TREEHERDER_DEBUG=False backend python -bb ./manage.py check --deploy --fail-level WARNING
         # Using `-bb` mode to surface BytesWarnings: https://docs.python.org/3.7/using/cmdline.html#cmdoption-b
-        - docker-compose run backend python -bb -m pytest tests/ --runslow --ignore=tests/selenium/
+        - docker-compose run backend bash -c "pytest --cov=./ --cov-report=xml tests/ --runslow --ignore=tests/selenium"
```
Comment (armenzg) [inline .travis.yml]: I don't believe you need `--cov=./` is that right?

File: docker/entrypoint.sh [hunk 19]
```diff
@@ -14,7 +14,4 @@ while ! nc -z "${DATABASE_HOST}" "${DATABASE_PORT}" &> /dev/null; do
     sleep 1;
 done
 
-# Initialize migrations and SETA
-./initialize_data.sh
```
[the remark this answers is not in the document]
Comment (armenzg) [reply docker/entrypoint.sh]: I also see such file deleted:
image

What did you do to resolve the merge conflicts?

File: package.json [hunk 20]
```diff
@@ -90,7 +90,7 @@
   "scripts": {
     "build": "node ./node_modules/webpack/bin/webpack.js --mode production",
     "build:dev": "node ./node_modules/webpack/bin/webpack.js --mode development",
-    "codecov": "node ./node_modules/codecov/bin/codecov && rm -rf coverage",
+    "codecov": "node ./node_modules/codecov/bin/codecov --token=995256be-3a1f-4d10-b181-e24f88603b6b && rm -rf coverage",
```
Comment (armenzg) [inline package.json]: We will need to remove all lines about tokens before merging.
I can take care of it before we merge.

File: requirements/dev.txt [hunk 21]
```diff
@@ -1,9 +1,16 @@
...
+    --hash=sha256:d649dc0bcace6fcdb446ae02b98798a856593b19b637c1b9af8edadf2b150bea
+
+pytest-cov==2.8.1 \
+    --hash=sha256:cdbdef4f870408ebdbfeb44e63e07eb18bb4619fae852f6e760645fa36172626
+
 pytest==5.3.5 \
     --hash=sha256:ff615c761e25eb25df19edddc0b970302d2a9091fbce0e7213298d85fb61fef6 \
     --hash=sha256:0d5fe9189a148acc3c3eb2ac8e1ac0742cb7618c084f3d228baaec0c254b318d
 
+
```
Comment (armenzg) [inline requirements/dev.txt]: Please remove this extra line.

Comment (shubhank-saxena) [reply requirements/dev.txt]: Oops. Drawbacks of using Vim! Let me rectify it

Comment (shubhank-saxena) [reply requirements/dev.txt]: @armenzg just a little doubt. How do you push the code to repo after 2F authentication is enabled?
Created my personal access token. But it's throwing error 403

File: package.json (hunk 20 above)
Comment (shubhank-saxena) [reply package.json]: Okay. I think coverage will work without the token too

File: docker/entrypoint.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (shubhank-saxena) [reply docker/entrypoint.sh]: Yes sir. You did mention to delete the file as it wasn't required

[the remark this answers is not in the document]
Comment (shubhank-saxena) [reply docker/entrypoint.sh]: > Was this an intentional change?

Yes after the file was deleted, needed to remove it from entrypoint.sh

File: .travis.yml (hunk 18 above)
Comment (shubhank-saxena) [reply .travis.yml]: It was part of the standard commands that I found everwhere. Let me try without it

Comment (shubhank-saxena) [reply .travis.yml]: Not forming the coverage.xml file if we omit ```--cov=./```

File: docker/entrypoint.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (shubhank-saxena) [reply docker/entrypoint.sh]: For resolving merge conflict I went with latest version of pytest. I deleted the older version and it's hashes

File: .travis.yml [hunk 22]
```diff
@@ -37,20 +37,26 @@ matrix:
...
       script:
         - docker-compose run backend ./runtests.sh
+      after_success:
+        - codecov -f coverage.xml -t 995256be-3a1f-4d10-b181-e24f88603b6b
 
     - env: python-tests-main
       language: minimal
       install:
         - docker-compose build
+        - pip install codecov --user
```
Comment (klahnakoski) [inline .travis.yml]: @armenzg are we able to run coverage locally?  I thought my suggestion of `/tests/coverage.sh` was good as an example, and to ensure coverage could be run locally.

Comment (armenzg) [reply .travis.yml]: I just saw this comment. I have a new branch to get this out the door so we can postpone anything that is missing to a follow up issue.

Support for running code coverage results locally is supported as part of this PR.

Now, submitting results from localhost to **codecov** requires setting up a personal codecov account, grabbing a token, installing `codecov` manually and then running the tests.

**Comment by armenzg** [comment]
I'm testing this on PR #5921
It will keep your authorship.
Thanks for the hard work!
